Testnet/NFTAuctionStore/NFTAuctionStore/NFTAuctionStore.cs

```diff
@@ -0,0 +1,199 @@
+
+        if (auction.HighestBid > 0)
+        {
+            //refund for previous bidder
+            var balance = GetRefund(auction.HighestBidder);
+
+            SetRefund(auction.HighestBidder, balance + auction.HighestBid);
+        }
+
+        auction.HighestBidder = Message.Sender;
+        auction.HighestBid = Message.Value;
+
+        SetAuctionInfo(contract, tokenId, auction);
+
+        Log(new HighestBidUpdatedLog { Contract = contract, TokenId = tokenId, Bidder = auction.HighestBidder, Bid = auction.HighestBid });
+    }
+
+    public bool Refund()
+    {
+        EnsureNotPayable();
+
+        var amount = GetRefund(Message.Sender);
+
+        Assert(amount > 0);
+
+        SetRefund(Message.Sender, 0);
+
+        var transfer = Transfer(this.Message.Sender, amount);
+
+        if (!transfer.Success)
+            SetRefund(Message.Sender, amount);
+
+        return transfer.Success;
+    }
+
+    public void AuctionEnd(Address contract, ulong tokenId)
+    {
+        EnsureNotPayable();
+
+        var auction = GetAuctionInfo(contract, tokenId);
+
+        Assert(Block.Number >= auction.EndBlock, "Auction is not ended yet.");
+
+        Assert(!auction.Ended,"Auction end already executed.");
+
+        auction.Ended = true;
+        SetAuctionInfo(contract, tokenId, auction);
+
+        if (auction.HighestBid > 0)
+        {
+            var result = Transfer(auction.Seller, auction.HighestBid);
+
+            Assert(result.Success, "Transfer failed.");
+
+            SafeTransferToken(contract, tokenId, Address, auction.HighestBidder);
+        }
+        else
+        {
+            SafeTransferToken(contract, tokenId, Address, auction.Seller);
+        }
+
+        Log(new AuctionEndedLog { Contract = contract, TokenId = tokenId, HighestBidder = auction.HighestBidder, HighestBid = auction.HighestBid });
+    }
+
+    private bool IsApprovedForAll(Address contract, Address tokenOwner)
+    {
+        var result = Call(contract, 0, "IsApprovedForAll", new object[] { tokenOwner, Message.Sender });
+
+        Assert(result.Success, "IsApprovedForAll method call failed.");
+
+        return result.ReturnValue is bool success && success;
+    }
+
+    private void TransferToken(Address contract, ulong tokenId, Address from, Address to)
+    {
+        var result = Call(contract, 0, "TransferFrom", new object[] { from, to, tokenId });
+
+        Assert(result.Success && result.ReturnValue is bool success && success, "The token transfer failed. Be sure sender is approved to transfer token.");
+    }
+
+    private void SafeTransferToken(Address contract, ulong tokenId, Address from, Address to)
+    {
+        var result = Call(contract, 0, "SafeTransferFrom", new object[] { from, to, tokenId });
+
+        Assert(result.Success && result.ReturnValue is bool success && success, "The token transfer failed.");
+    }
```

Comment:
Is there a difference between these two methods? Obviously method name is different but from a reviewers perspective, I have no idea.

Reply:
SafeTransferFrom works same as TransferFrom on NFT contract but in case of To is another smart contract then SafeTransferFrom does a call OnReceived method as notification on To(contract)